File: 03-operator-cpp/Fraction.cpp
#include <iostream>
#include <numeric>
#include <stdexcept>

#include "Fraction.h"

Fraction::Fraction(int numenator, int denuminator)
{
    setNumenator(numenator);
    setDenuminator(denuminator);
    normalize();
}

void Fraction::setNumenator(int value)
{
    m_numenator = value;
}

void Fraction::setDenuminator(int value)
{
    if (value == 0)
    {
        throw std::invalid_argument("Знаменатель не может быть равен нулю");
    }
    m_denuminator = value;
}

void Fraction::normalize()
{
    const int n1 = getNumenator();
    const int d1 = getDenuminator();
    const int gcd = std::gcd(n1, d1);  // c++17
    setNumenator(n1 / gcd);
    setDenuminator(d1 / gcd);
}

Fraction Fraction::operator+(const Fraction& rhs) const
{
    const int n1 = getNumenator();
    const int d1 = getDenuminator();
    const int n2 = rhs.getNumenator();
    const int d2 = rhs.getDenuminator();

    const int n11 = n1 * d2;
    const int n22 = n2 * d1;
    const int d11 = d1 * d2;

    return Fraction(n11 + n22, d11);
}

Fraction& Fraction::operator++()
{
    const int n1 = getNumenator();
    const int d1 = getDenuminator();
    setNumenator(n1 + d1);  // +1
    normalize();
    return *this;
}

Fraction Fraction::operator++(int)
{
    // копируется объект
    Fraction old = *this;
    // переиспользуется перегрузка префиксного инкремента
    ++*this;
    return old;
}

std::ostream& operator<<(std::ostream& os, const Fraction& obj)
{
    os << obj.getNumenator() << '/' << obj.getDenuminator();
    return os;
}

File: 03-operator-dotnet/Fraction.cs
using System.Numerics;

namespace _03_operator_dotnet
{
    /// <summary>
    /// Дробь
    /// </summary>
    public class Fraction
    {
        /// <summary>
        /// Конструктор дроби
        /// </summary>
        /// <param name="numenator">числитель</param>
        /// <param name="denuminator">знаменатель</param>
        public Fraction(BigInteger numenator, BigInteger denuminator)
        {
            Numenator = numenator;
            Denuminator = denuminator;
            Normalize();
        }

        /// <summary>
        /// Перегрузка оператора "+"
        /// </summary>
        /// <param name="lhs">левый операнд</param>
        /// <param name="rhs">правый операнд</param>
        /// <returns>Новый объект дроби</returns>
        public static Fraction operator +(Fraction lhs, Fraction rhs)
        {
            BigInteger numenator = lhs.Numenator * rhs.Denuminator + rhs.Numenator * lhs.Denuminator;
            BigInteger denuminator = lhs.Denuminator * rhs.Denuminator;
            return new Fraction(numenator, denuminator);
        }

        /// <summary>
        /// Перегрузка оператора "++"
        /// </summary>
        public static Fraction operator ++(Fraction obj)
        {
            return new Fraction(obj.Numenator + obj.Denuminator, obj.Denuminator);  // +1
        }

        public override string ToString()
        {
            return Numenator + "/" + Denuminator;
        }

        /// <summary>
        /// Числитель
        /// </summary>
        public BigInteger Numenator { get; private set; }

        /// <summary>
        /// Знаменатель
        /// </summary>
        public BigInteger Denuminator
        {
            get { return m_denuminator; }
            set { ValidateDenuminator(value); m_denuminator = value; }
        }

        private BigInteger m_denuminator;

        private static void ValidateDenuminator(BigInteger denuminator)
        {
            if (denuminator == 0)
            {
                throw new ArgumentException("denuminator");
            }
        }

        /// <summary>
        /// Нормализовать число
        /// </summary>
        private void Normalize()
        {
            BigInteger gcd = BigInteger.GreatestCommonDivisor(Numenator, Denuminator);
            Numenator /= gcd;
            Denuminator /= gcd;
        }
    }
}
